addons/csi-vault-secret-provider/README.md

```diff
@@ -0,0 +1,154 @@
+# Vault CSI driver
+
+See more: https://github.com/hashicorp/vault-csi-provider
+
+basic YAML generated by:
+
+```
+helm template vault hashicorp/vault \
+    --namespace=kube-system \
+    --values=generate-values \
+    > manifest.yaml
+```
+
+## KubeOne cluster init
+
+If you are only creating the cluster, enable corresponding addons in the KubeOne
+manifest:
+
+```yaml
+apiVersion: kubeone.k8c.io/v1beta2
+kind: KubeOneCluster
+versions:
+  kubernetes: 1.23.6
+    
+containerRuntime:
+  containerd: {}
+
+addons:
+  enable: true
+  addons:
+  - name: secrets-store-csi-driver
+  - name: csi-vault-secret-provider
+```
+
+And init the cluster as usually.
+
+## Configure Vault
+
+Example configuration for AWS access.
+You have to setup shell variables:
+* `AWS_ACCESS_KEY_ID`
+* `AWS_SECRET_ACCESS_KEY`
+* `SERVICE_ACCOUNT_TOKEN`
+  A token from service account with `ClusterRole` binding to `system:auth-delegator`
+* `KUBERNETES_API_ENDPOINT`
+
+and `ca.crt` file of your kubernetes CA Cert.
+
+```shell
+vault kv put secret/cloud-credentials \
+    AWS_ACCESS_KEY_ID="${AWS_ACCESS_KEY_ID}" \
+    AWS_SECRET_ACCESS_KEY="${AWS_SECRET_ACCESS_KEY}"
+
+vault auth enable kubernetes
+
+vault write auth/kubernetes/config \
+    token_reviewer_jwt="${SERVICE_ACCOUNT_TOKEN}" \
+    kubernetes_host="https://${KUBERNETES_API_ENDPOINT}:6443" \
+    kubernetes_ca_cert=@ca.crt
+
+vault policy write kubeone - <<EOF
+path "secret/data/cloud-credentials" {
+  capabilities = ["read"]
+}
+EOF
+
+# following service accounts are used by the deployments that need access to
+# credentials
+vault write auth/kubernetes/role/kubeone \
+    bound_service_account_names=machine-controller \
+    bound_service_account_names=machine-controller-webhook \
+    bound_service_account_names=operating-system-manager \
+    bound_service_account_names=operating-system-manager-webhook \
+    bound_service_account_names=cloud-controller-manager \
+    bound_service_account_names=ebs-csi-controller-sa \
+    bound_service_account_names=ebs-csi-node-sa \
+    bound_service_account_namespaces=kube-system \
+    policies=kubeone \
+    ttl=20m
```

Comment:
I think TTL is the "time to live" for tokens to access vault that are generated by the service accounts.

Reply:
Indeed this is Time To Live, for the vault token. It's specifically short-lived so, to quickly invalidate (expire) it in case it leaks.

Reply:
What's going to happen if you delete/rotate that pod after TTL? Is it going to fail to get the secret/credentials?

Reply:
new pod will be started, since it has same kubernetes identity (service-account/namespace) it will receive new token. I suppose every pod replica will receive own short-lived token.

Reply:
Can we document this in the README?

Reply:
Will link to the https://www.vaultproject.io/docs/auth/kubernetes make you happy?

Reply:
Yes, it looks like a good starter.